core/src/OC/mimeType.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import { generateUrl } from '@nextcloud/router'
+
+const iconCache = new Map()
+
+/**
+ * Return the url to icon of the given mimeType
+ *
+ * @param {string} mimeType The mimeType to get the icon for
+ * @return {string} Url to the icon for mimeType
+ */
+export function getIconUrl(mimeType) {
+	if (typeof mimeType === 'undefined') {
+		return undefined
+	}
+
+	while (mimeType in window.OC.MimeTypeList.aliases) {
+		mimeType = window.OC.MimeTypeList.aliases[mimeType]
+	}
+
+	if (!iconCache.has(mimeType)) {
+		// First try to get the correct icon from the current theme
+		let gotIcon = null
+		let path = ''
+		if (OC.theme.folder !== '' && Array.isArray(OC.MimeTypeList.themes[OC.theme.folder])) {
+			path = generateUrl('/themes/' + window.OC.theme.folder + '/core/img/filetypes/')
+			const icon = getMimeTypeIcon(mimeType, window.OC.MimeTypeList.themes[OC.theme.folder])
+
+			if (icon !== null) {
+				gotIcon = true
+				path += icon
+			}
+		}
+		if (window.OCA.Theming && gotIcon === null) {
+			path = generateUrl('/apps/theming/img/core/filetypes/')
+			path += getMimeTypeIcon(mimeType, window.OC.MimeTypeList.files)
+			gotIcon = true
+		}
+
+		// If we do not yet have an icon fall back to the default
+		if (gotIcon === null) {
+			path = generateUrl('/core/img/filetypes/')
+			path += getMimeTypeIcon(mimeType, window.OC.MimeTypeList.files)
+		}
```

Comment:
It is not only moving the function but also changing it
(using suggestion just to show a diff)
```suggestion
		if (OCA.Theming && gotIcon === null) {
			path = OC.generateUrl('/apps/theming/img/core/filetypes/')
			path += OC.MimeType._getFile(mimeType, OC.MimeTypeList.files)
			gotIcon = true
		}

		// If we do not yet have an icon fall back to the default
		if (gotIcon === null) {
			path = OC.getRootPath() + '/core/img/filetypes/'
			path += OC.MimeType._getFile(mimeType, OC.MimeTypeList.files)
		}
```

But `generateUrl` cannot be used here.

1. `/apps/theming/img/` - is a PHP route managed by a controller
2. '/core/img/filetypes/' - is a static file path

In the first case we need to include `index.php` when no rewrite. But in the second adding `index.php` breaks the path to the icon. 

`/index.php/core/img/filetypes/file.svg` is not correct.

It must be either `imagePath` or `generateFilePath`, or `getRootPath` like in the original version. But it cannot be `generateUrl`

Reply:
cc @susnux